### ISSUE TITLE
dbcsr_tensor_unittest: fails with Cray and Intel compiler

### ISSUE THREAD

**Issue by dev-zero** [cp2k/dbcsr#201]
**Describe the bug**

`dbcsr_tensor_unittest` fails on Piz Daint with both Cray and Intel 17 compilers. While the Cray compiler doesn't yield any useful information, I get the following when built with Intel:

```
12/17 Test #12: dbcsr_tensor_unittest .................................***Failed    5.73 sec

 

--------------------------------------------------------------------------------

Testing matrix representations of tensor rank 2

--------------------------------------------------------------------------------

Block sizes:

   Dim 1:  3  5  1 23  2  3  1  6  3  8  2  3  5  1 

   Dim 2:  4  2  5  3  1  5 13  5  2  4  5  6  7  2  3  1  2  6  9 12 21 

Non-zero blocks:

   Block  1: (  1  1 )

   Block  2: (  1  3 )

   Block  3: (  1 11 )

   Block  4: (  2 15 )

   Block  5: (  4  4 )

   Block  6: (  4 17 )

   Block  7: (  7 21 )

   Block  8: ( 10  6 )

   Block  9: ( 10  9 )

   Block 10: ( 10 13 )

   Block 11: ( 10 19 )

   Block 12: ( 13  7 )

 

Reference map: ( 1 | 2 )

 

Test   1: ( 1 | 2 )

   Reference distribution:

      Dist vec 1:  1  1  0  0  1  0  0  0  1  1  0  1  0  1 

      Dist vec 2:  0  1  1  1  0  0  0  1  1  1  1  0  0  0  1  0  1  1  0  0  1 

   Test distribution:

      Dist vec 1:  0  0  0  1  1  1  0  1  1  1  0  1  0  1 

      Dist vec 2:  1  0  1  1  0  0  1  1  0  1  1  0  1  0  1  1  1  0  1  1  0 

forrtl: severe (188): An assignment was made from an object of one size to an object of a different size that cannot be deallocated

Image              PC                Routine            Line        Source             

libifcore.so.5     00002AAAB7036A9C  for_alloc_assign_     Unknown  Unknown

dbcsr_tensor_unit  000000000074605B  Unknown               Unknown  Unknown

dbcsr_tensor_unit  0000000000661410  Unknown               Unknown  Unknown

dbcsr_tensor_unit  0000000000633284  Unknown               Unknown  Unknown

dbcsr_tensor_unit  000000000040E8FA  Unknown               Unknown  Unknown

dbcsr_tensor_unit  000000000040DF4E  Unknown               Unknown  Unknown

libc-2.22.so       00002AAAB8A27725  __libc_start_main     Unknown  Unknown

dbcsr_tensor_unit  000000000040DE49  Unknown               Unknown  Unknown

forrtl: severe (188): An assignment was made from an object of one size to an object of a different size that cannot be deallocated

Image              PC                Routine            Line        Source             

libifcore.so.5     00002AAAB7036A9C  for_alloc_assign_     Unknown  Unknown

dbcsr_tensor_unit  000000000074605B  Unknown               Unknown  Unknown

dbcsr_tensor_unit  0000000000661410  Unknown               Unknown  Unknown

dbcsr_tensor_unit  0000000000633284  Unknown               Unknown  Unknown

dbcsr_tensor_unit  000000000040E8FA  Unknown               Unknown  Unknown

dbcsr_tensor_unit  000000000040DF4E  Unknown               Unknown  Unknown

libc-2.22.so       00002AAAB8A27725  __libc_start_main     Unknown  Unknown

dbcsr_tensor_unit  000000000040DE49  Unknown               Unknown  Unknown

forrtl: severe (188): An assignment was made from an object of one size to an object of a different size that cannot be deallocated

Image              PC                Routine            Line        Source             

libifcore.so.5     00002AAAB7036A9C  for_alloc_assign_     Unknown  Unknown

dbcsr_tensor_unit  000000000074605B  Unknown               Unknown  Unknown

dbcsr_tensor_unit  0000000000661410  Unknown               Unknown  Unknown

dbcsr_tensor_unit  0000000000633284  Unknown               Unknown  Unknown

dbcsr_tensor_unit  000000000040E8FA  Unknown               Unknown  Unknown

dbcsr_tensor_unit  000000000040DF4E  Unknown               Unknown  Unknown

libc-2.22.so       00002AAAB8A27725  __libc_start_main     Unknown  Unknown

dbcsr_tensor_unit  000000000040DE49  Unknown               Unknown  Unknown

forrtl: severe (188): An assignment was made from an object of one size to an object of a different size that cannot be deallocated

Image              PC                Routine            Line        Source             

libifcore.so.5     00002AAAB7036A9C  for_alloc_assign_     Unknown  Unknown

dbcsr_tensor_unit  000000000074605B  Unknown               Unknown  Unknown

dbcsr_tensor_unit  0000000000661410  Unknown               Unknown  Unknown

dbcsr_tensor_unit  0000000000633284  Unknown               Unknown  Unknown

dbcsr_tensor_unit  000000000040E8FA  Unknown               Unknown  Unknown

dbcsr_tensor_unit  000000000040DF4E  Unknown               Unknown  Unknown

libc-2.22.so       00002AAAB8A27725  __libc_start_main     Unknown  Unknown

dbcsr_tensor_unit  000000000040DE49  Unknown               Unknown  Unknown

```

**To Reproduce**
https://github.com/cp2k/dbcsr/blob/05fcdbb276072a55788a628e94f424341426577d/.ci/daint.cscs.ch/intel.build.sh
https://github.com/cp2k/dbcsr/blob/05fcdbb276072a55788a628e94f424341426577d/.ci/daint.cscs.ch/intel.test.sh

**Comment by dev-zero** [comment]
with `-traceback` there is some more info:

```
forrtl: severe (188): An assignment was made from an object of one size to an object of a different size that cannot be deallocated
Image              PC                Routine            Line        Source             
libifcore.so.5     00002AAAB7036A9C  for_alloc_assign_     Unknown  Unknown
dbcsr_tensor_unit  0000000000745C2B  dbcsr_tas_base_mp         210  dbcsr_tas_base.F
dbcsr_tensor_unit  0000000000660E80  dbcsr_tensor_type         768  dbcsr_tensor_types.F
dbcsr_tensor_unit  0000000000632D34  dbcsr_tensor_test         441  dbcsr_tensor_test.F
dbcsr_tensor_unit  000000000040E8FA  MAIN__                    105  dbcsr_tensor_unittest.F
dbcsr_tensor_unit  000000000040DF4E  Unknown               Unknown  Unknown
libc-2.22.so       00002AAAB8C2B725  __libc_start_main     Unknown  Unknown
dbcsr_tensor_unit  000000000040DE49  Unknown               Unknown  Unknown
forrtl: severe (188): An assignment was made from an object of one size to an object of a different size that cannot be deallocated
Image              PC                Routine            Line        Source             
libifcore.so.5     00002AAAB7036A9C  for_alloc_assign_     Unknown  Unknown
dbcsr_tensor_unit  0000000000745C2B  dbcsr_tas_base_mp         210  dbcsr_tas_base.F
dbcsr_tensor_unit  0000000000660E80  dbcsr_tensor_type         768  dbcsr_tensor_types.F
dbcsr_tensor_unit  0000000000632D34  dbcsr_tensor_test         441  dbcsr_tensor_test.F
dbcsr_tensor_unit  000000000040E8FA  MAIN__                    105  dbcsr_tensor_unittest.F
dbcsr_tensor_unit  000000000040DF4E  Unknown               Unknown  Unknown
libc-2.22.so       00002AAAB8C2B725  __libc_start_main     Unknown  Unknown
dbcsr_tensor_unit  000000000040DE49  Unknown               Unknown  Unknown
forrtl: severe (188): An assignment was made from an object of one size to an object of a different size that cannot be deallocated
Image              PC                Routine            Line        Source             
libifcore.so.5     00002AAAB7036A9C  for_alloc_assign_     Unknown  Unknown
dbcsr_tensor_unit  0000000000745C2B  dbcsr_tas_base_mp         210  dbcsr_tas_base.F
dbcsr_tensor_unit  0000000000660E80  dbcsr_tensor_type         768  dbcsr_tensor_types.F
dbcsr_tensor_unit  0000000000632D34  dbcsr_tensor_test         441  dbcsr_tensor_test.F
dbcsr_tensor_unit  000000000040E8FA  MAIN__                    105  dbcsr_tensor_unittest.F
dbcsr_tensor_unit  000000000040DF4E  Unknown               Unknown  Unknown
libc-2.22.so       00002AAAB8C2B725  __libc_start_main     Unknown  Unknown
dbcsr_tensor_unit  000000000040DE49  Unknown               Unknown  Unknown
forrtl: severe (188): An assignment was made from an object of one size to an object of a different size that cannot be deallocated
Image              PC                Routine            Line        Source             
libifcore.so.5     00002AAAB7036A9C  for_alloc_assign_     Unknown  Unknown
dbcsr_tensor_unit  0000000000745C2B  dbcsr_tas_base_mp         210  dbcsr_tas_base.F
dbcsr_tensor_unit  0000000000660E80  dbcsr_tensor_type         768  dbcsr_tensor_types.F
dbcsr_tensor_unit  0000000000632D34  dbcsr_tensor_test         441  dbcsr_tensor_test.F
dbcsr_tensor_unit  000000000040E8FA  MAIN__                    105  dbcsr_tensor_unittest.F
dbcsr_tensor_unit  000000000040DF4E  Unknown               Unknown  Unknown
libc-2.22.so       00002AAAB8C2B725  __libc_start_main     Unknown  Unknown
dbcsr_tensor_unit  000000000040DE49  Unknown               Unknown  Unknown
srun: error: nid02585: tasks 0-3: Exited with exit code 188
srun: Terminating job step 15018403.11
```

**Comment by pseewald** [comment]
Thank you, I'll have a look and try to fix it.

**Comment by dev-zero** [comment]
... after pre-processing, line 210 in `dbcsr_tas_base.F` is: `CALL dbcsr_tas_copy_distribution(dist, matrix%dist, own_dist)`

**Comment by dev-zero** [comment]
Interestingly, when building with `-free -stand f08 -fpp -traceback -O0 -debug -check all` I don't get the same error, only warnings about array temporaries being created:

```
forrtl: warning (406): fort: (1): In call to ALLOCATE_1D_I, an array temporary was created for argument #2
forrtl: warning (406): fort: (1): In call to ALLOCATE_2D_R_DP, an array temporary was created for argument #2
forrtl: warning (406): fort: (1): In call to BLOCK_BUFFER_ADD_AREA_BLOCK, an array temporary was created for argument #2
forrtl: warning (406): fort: (1): In call to CALL_REC_SORT_INDEX, an array temporary was created for argument #4
forrtl: warning (406): fort: (1): In call to COUNT_BINS, an array temporary was created for argument #2
forrtl: warning (406): fort: (1): In call to DBCSR_SCALE_ANYTYPE, an array temporary was created for argument #3
forrtl: warning (406): fort: (1): In call to DBCSR_T_GET_STORED_COORDINATES, an array temporary was created for argument #2
forrtl: warning (406): fort: (1): In call to GET_ARRAY_ELEMENTS, an array temporary was created for argument #2
forrtl: warning (406): fort: (1): In call to GET_ND_INDICES, an array temporary was created for argument #2
forrtl: warning (406): fort: (1): In call to I/O Write routine, an array temporary was created for argument #2
forrtl: warning (406): fort: (1): In call to LOCAL_FILTER, an array temporary was created for argument #3
forrtl: warning (406): fort: (1): In call to MPI_WAITALL_INTERNAL, an array temporary was created for argument #2
```

**Comment by pseewald** [comment]
I could not find anything wrong in the code. Judging from the error message `An assignment was made from an object of one size to an object of a different size that cannot be deallocated` the problem may or may not be related to `matrix` being `intent(out)` in the subroutine that calls `dbcsr_tas_copy_distribution(dist, matrix%dist, own_dist)` (see also [this](https://software.intel.com/en-us/forums/intel-fortran-compiler-for-linux-and-mac-os-x/topic/799997)).

I suggest we try a recent version of Intel compiler before investigating further.

**Comment by alazzaro** [comment]
I think this issue is fixed via #290 , @pseewald is this the case? If so, please close it.
